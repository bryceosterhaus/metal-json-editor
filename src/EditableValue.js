import Component, {Config} from 'metal-jsx';

import {TYPE_COLORS} from './util';

class EditableValue extends Component {
	getTypeStyle(value) {
		const {readOnly} = this.props;

		let type = '';

		if (value === null) {
			type = 'null';
		} else if (readOnly) {
			type = 'readOnly';
		} else {
			type = typeof value;
		}

		return `color: ${TYPE_COLORS[type]};`;
	}

	handleValueEdit() {
		if (!this.props.readOnly) {
			this.setState(
				{
					editing: true
				},
				() => this.refs.input.focus()
			);
		}
	}

	submitValue({target}) {
		const valType = typeof this.props.value;

		this.state.editing = false;

		let newVal = target.value;

		if (valType === 'boolean') {
			newVal = newVal === 'true' ? true : false;
		} else if (valType === 'number') {
			newVal = Number(newVal);
		}

		if (newVal !== this.props.value) {
			this.props.onChange(this.props.locator, newVal);
		}
	}

	render() {
		let {readOnly, value} = this.props;
		const {editing} = this.state;

		if (value && value.__metal_devtools_read_only) {
			value = value.value;
		}

		return (
			<span
				style={this.getTypeStyle(value)}
				onClick={this.handleValueEdit.bind(this)}
				title={readOnly ? 'Read Only' : 'Click to Edit'}
			>
				{!editing &&
					value !== undefined &&
					<span style="white-space: pre;">{value + ''}</span>}

				{!editing &&
					(value === undefined || value === '') &&
					<span>
						""
					</span>}

				{editing &&
					<input
						ref="input"
						onBlur={this.submitValue.bind(this)}
						value={value}
					/>}
			</span>
		);
	}
}

EditableValue.PROPS = {
	name: Config.string(),
	onChange: Config.func(),
	readOnly: Config.bool().value(false),
	value: Config.any()
};

EditableValue.STATE = {
	editing: Config.value(false)
};

export default EditableValue;
