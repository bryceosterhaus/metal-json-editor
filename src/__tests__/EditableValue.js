import EditableValue from '../EditableValue';

describe(
	'EditableValue',
	() => {
		let component;

		afterEach(
			() => {
				if (component) {
					component.dispose();
				}
			}
		);

		it(
			'renders',
			() => {
				component = new EditableValue();

				expect(component).toMatchSnapshot();
			}
		);

		it(
			'should render with editing enabled',
			() => {
				component = new EditableValue(
					{
						value: 1
					}
				);

				component.state.editing = true;

				jest.runAllTimers();

				expect(component).toMatchSnapshot();
			}
		);

		it(
			'should return color value',
			() => {
				component = new EditableValue();

				expect(component.getTypeStyle('')).toMatchSnapshot();
				expect(component.getTypeStyle(null)).toMatchSnapshot();
				expect(component.getTypeStyle(3)).toMatchSnapshot();

				component.props.readOnly = true;

				expect(component.getTypeStyle('Function()')).toMatchSnapshot();
			}
		);

		it(
			'should update state',
			() => {
				component = new EditableValue();

				component.refs.input = {
					focus: jest.fn()
				};

				component.handleValueEdit();

				jest.runAllTimers();

				expect(component.state).toMatchSnapshot();
			}
		);

		it(
			'should call onChange with new number value',
			() => {
				const spy = jest.fn();

				component = new EditableValue(
					{
						onChange: spy,
						value: 1
					}
				);

				component.submitValue(
					{
						target: {value: 2}
					}
				);

				jest.runAllTimers();

				expect(spy).toHaveBeenCalledTimes(1);
				expect(component.state).toMatchSnapshot();

				component.props.value = true;

				component.submitValue(
					{
						target: {value: 'false'}
					}
				);

				jest.runAllTimers();

				expect(spy).toHaveBeenCalledTimes(2);
				expect(component.state).toMatchSnapshot();

				component.props.value = false;

				component.submitValue(
					{
						target: {value: 'true'}
					}
				);

				jest.runAllTimers();

				expect(spy).toHaveBeenCalledTimes(3);
				expect(component.state).toMatchSnapshot();
			}
		);

		it(
			'should render in readOnly mode',
			() => {
				component = new EditableValue(
					{
						readOnly: true,
						value: {
							__metal_devtools_read_only: true,
							value: 'Function()'
						}
					}
				);

				expect(component).toMatchSnapshot();
			}
		);
	}
);
