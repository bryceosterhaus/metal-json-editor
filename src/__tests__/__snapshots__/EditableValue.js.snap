// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableValue renders 1`] = `
<span style="" title="Click to Edit">
  <span key="_metal_jsx_2"> "" </span>
</span>
`;

exports[`EditableValue should call onChange with new number value 1`] = `
Object {
  "editing": false,
}
`;

exports[`EditableValue should call onChange with new number value 2`] = `
Object {
  "editing": false,
}
`;

exports[`EditableValue should call onChange with new number value 3`] = `
Object {
  "editing": false,
}
`;

exports[`EditableValue should render in readOnly mode 1`] = `
<span style="color: rgb(17, 119, 0);" title="Read Only">
  <span style="white-space: pre;" key="_metal_jsx_1">Function()</span>
</span>
`;

exports[`EditableValue should render with editing enabled 1`] = `
<span style="color: rgb(28, 0, 207);" title="Click to Edit">
  <input value="1" ref="input" key="_metal_jsx_3">
</span>
`;

exports[`EditableValue should return color value 1`] = `"color: #C41A16;"`;

exports[`EditableValue should return color value 2`] = `"color: #FF8C00;"`;

exports[`EditableValue should return color value 3`] = `"color: #1C00CF;"`;

exports[`EditableValue should return color value 4`] = `"color: #117700;"`;

exports[`EditableValue should update state 1`] = `
Object {
  "editing": true,
}
`;
